libraries/bot-connector/src/main/java/com/microsoft/bot/connector/authentication/OAuthClient.java

```diff
@@ -141,85 +177,124 @@ else if (statusCode == HTTP_NOT_FOUND) {
-
+                .withConnectionName(connectionName)
+                .withConversation(new ConversationReference()
+                        .withActivityId(activity.id())
+                        .withBot(activity.recipient())
+                        .withChannelId(activity.channelId())
+                        .withConversation(activity.conversation())
+                        .withServiceUrl(activity.serviceUrl())
+                        .withUser(activity.from()))
+                .withMsAppId((creds == null) ? null : creds.microsoftAppId());
 
         String serializedState = this.mapper.writeValueAsString(tokenExchangeState);
-        String encoded = Base64.getEncoder().encodeToString(serializedState.getBytes(StandardCharsets.UTF_8));
 
         // Construct URL
-
-        HashMap <String, String> qstrings = new HashMap<>();
+        String encoded = Base64.getEncoder().encodeToString(serializedState.getBytes(StandardCharsets.UTF_8));
+        HashMap<String, String> qstrings = new HashMap<>();
         qstrings.put("state", encoded);
-        String strUri = String.format("%sapi/botsignin/getsigninurl", this.uri);
-        URI tokenUrl = MakeUri(strUri, qstrings);
-
-        // add botframework api service url to the list of trusted service url's for these app credentials.
-        MicrosoftAppCredentials.trustServiceUrl(tokenUrl);
-
-        // Set Credentials and make call
-        MicrosoftAppCredentials appCredentials = (MicrosoftAppCredentials) client.restClient().credentials();
-        Response httpResponse = await(appCredentials.ProcessHttpRequestAsync(true, "GET", tokenUrl.toString()));
 
-        int statusCode = httpResponse.code();
-        if (statusCode == HTTP_OK) {
-            return completedFuture(httpResponse.body().string());
-        }
-        return completedFuture(null);
+        String strUri = String.format("%sapi/botsignin/getsigninurl", this.uri);
+        final URI tokenUrl = MakeUri(strUri, qstrings);
+
+        return CompletableFuture.supplyAsync(() -> {
+
+            // add botframework api service url to the list of trusted service url's for these app credentials.
+            MicrosoftAppCredentials.trustServiceUrl(tokenUrl);
+
+
+            // Later: Use client in clientimpl?
+            OkHttpClient client = new OkHttpClient.Builder()
+                    .addInterceptor(new MicrosoftAppCredentialsInterceptor(creds))
+                    .build();
+
+            Request request = new Request.Builder()
+                    .url(tokenUrl.toString())
+                    .header("User-Agent", UserAgent.value())
+                    .build();
+
+            Response response = null;
+            try {
+                response = client.newCall(request).execute();
+                int statusCode = response.code();
+                if (statusCode == HTTP_OK)
+                    return response.body().string();
+            } catch (IOException e) {
+                e.printStackTrace();
```

Comment:
eat exception intended?

Reply:
Yeah, hard to get context - these are retrieving strings and tokens and stuff.  If it fails, just returns null or false in one case.